src/chk/chk_engine.c

```diff
@@ -52,14 +52,37 @@ struct chk_query_xstream_args {
 	struct chk_query_target		 cqxa_target;
 };
 
-static inline bool
-chk_engine_on_leader(d_rank_t leader)
+static void
+chk_destroy_pool_tree(struct chk_instance *ins)
 {
-	return dss_self_rank() == leader;
+	struct chk_pool_rec	*cpr;
+	struct chk_pool_rec	*tmp;
+
+	/*
+	 * Take reference on each pool record to guarantee that the pool list will not be
+	 * broken when we traverse the pool list even if there is yield during the travel.
+	 */
+	d_list_for_each_entry(cpr, &ins->ci_pool_list, cpr_link)
+		chk_pool_get(cpr);
+
+	/*
+	 * Once the pool record is deleted from the tree, the initial reference held when
+	 * created will be released: if it is current ULT delete the record from the tree,
+	 * then it will be via chk_pool_free()->chk_pool_put(). Otherwise, if it has been
+	 * deleted from the tree by others, then related logic wii call chk_pool_put().
+	 */
+	chk_destroy_tree(&ins->ci_pool_hdl, &ins->ci_pool_btr);
+
+	d_list_for_each_entry_safe(cpr, tmp, &ins->ci_pool_list, cpr_link) {
+		chk_pool_wait(cpr);
+		chk_pool_shutdown(cpr);
+		/* Release the reference held just above. */
+		chk_pool_put(cpr);
```

Comment:
Just curious who will delete the cpr_link?

Reply:
The last one called the chk_pool_put(), it is usually chk_engine_exit().